### PR TITLE
Add core search internal endpoint

### PR DESCRIPTION
**Pull request by coszio**
We want to support different inputs other than just a single query vector, so ultimately we had two options: 
1. Create a custom search internal endpoint, to handle other types of searches different from normal kNN.
2. Create a core search internal endpoint, to handle ALL types of searches, connecting directly with the already existing `CoreSearchRequest`

The first one sounds fine until I realized we'll have to duplicate a lot of code to make it work, which means more code to maintain.

At the end I chose to do the second approach, and plan to eventually deprecate the internal normal search endpoint and route everything through the `CoreSearchBatch` internal service. The plan is:
- For next release, we keep both endpoints so that search is always routed through `SearchBatch` while new recommend will be routed through `CoreSearchBatch`.
- In next next release, we'll safely remove `SearchBatch` internal endpoint and cleanup the duplicated code, routing everything to `CoreSearchBatch`.

### All Submissions:

* [x] Contributions should target the `dev` branch. Did you create your branch from `dev`?
* [x] Have you followed the guidelines in our Contributing document?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?

### New Feature Submissions:

1. [x] Does your submission pass tests?
4. [x] Have you formatted your code locally using `cargo +nightly fmt --all` command prior to submission?
5. [x] Have you checked your code using `cargo clippy --all --all-features` command?

### Changes to Core Features:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you successfully ran tests with your changes locally?

### CODE REVIEW

**Comment by coszio** [comment]
Just realized we need to actually enable normal search in `CoreSearchBatch` in order to enable zero-downtime migration, even if it won't be used in next release. Pushing these changes soon.

File: lib/collection/src/collection.rs [hunk 1]
```diff
@@ -916,14 +996,41 @@ impl Collection {
             try_join_all(all_searches).await?
         };
 
+        // TODO(luis): see if we can avoid this clone by removing the Arc
```
Comment (agourlay) [inline lib/collection/src/collection.rs:999]: Want to keep the TODO for later?

Comment (coszio) [reply lib/collection/src/collection.rs:999]: Whoops, must've reappeared when rebasing

File: openapi/tests/openapi_integration/helpers/collection_setup.py [hunk 2]
```diff
@@ -1,4 +1,3 @@
-import os
```
Comment (agourlay) [inline openapi/tests/openapi_integration/helpers/collection_setup.py:1]: I thought you fixed all those in a dedicated PR.
Do you a rebase maybe here?

Comment (coszio) [reply openapi/tests/openapi_integration/helpers/collection_setup.py]: 🤦 rebasing

**Comment by coszio** [comment]
To make the deprecation plan easier to understand, here is a diagram from the point of view of normal search during migrations:

imagen

The last update can be just a patch bump because CoreSearchPoints will be available in both 1.6 and 1.7

File: lib/collection/src/collection.rs [hunk 3]
```diff
@@ -840,6 +841,85 @@ impl Collection {
         }
     }
 
+    pub async fn core_search_batch(
```
Comment (timvisee) [inline lib/collection/src/collection.rs:844]: We still have `search_batch`.

Can we remove it's body, transform the request parameters, and call into this function to remove duplicate code?

File: lib/collection/src/collection.rs [hunk 4]
```diff
@@ -858,14 +938,43 @@ impl Collection {
...
 
+        let request = Arc::into_inner(request)
+            .expect("We have already dropped all of the Arc clones at this point")
+            .into();
+
+        self.merge_from_shards(all_searches_res, request, shard_selection)
+            .await
+    }
+
+    pub async fn _core_search_batch(
```
Comment (timvisee) [inline lib/collection/src/collection.rs:949]: Same here. Maybe we can use this for `_search_batch`?

File: src/tonic/api/points_internal_api.rs [hunk 5]
```diff
@@ -214,7 +215,34 @@ impl PointsInternal for PointsInternalService {
...
             collection_name,
             search_points,
-            None, // *Have* to be `None`!
+            None, // *Has* to be `None`!
+            shard_id,
+        )
+        .await
+    }
+
+    async fn core_search_batch(
```
Comment (timvisee) [inline src/tonic/api/points_internal_api.rs:224]: Another potential candidate? (`search_batch`)

File: src/tonic/api/points_common.rs [hunk 6]
```diff
@@ -760,6 +758,46 @@ pub async fn search_batch(
     Ok(Response::new(response))
 }
 
+pub async fn core_search_batch(
```
Comment (timvisee) [inline src/tonic/api/points_common.rs:761]: Another potential candidate? (`search_batch`)

File: lib/collection/src/shards/replica_set.rs [hunk 7]
```diff
@@ -1755,6 +1755,23 @@ impl ShardReplicaSet {
         .await
     }
 
+    pub async fn custom_search(
```
Comment (timvisee) [inline lib/collection/src/shards/replica_set.rs:1758]: Would `core_search` make more sense?

```suggestion
    pub async fn core_search(
```

File: lib/collection/src/shards/remote_shard.rs [hunk 8]
```diff
@@ -530,6 +532,58 @@ impl ShardOperation for RemoteShard {
         result
     }
 
+    async fn core_search(
```
Comment (timvisee) [inline lib/collection/src/shards/remote_shard.rs:535]: Another candidate? (`search`)

**Review by timvisee — APPROVED** [review]
Nice!

I tagged some instances where we may be able to reduce duplicate code. You may very well have looked at this already discovering that this wasn't easily possible, but this is hard to see while reviewing. I'll give this a shot myself and resolve cases where this isn't feasible.

Thanks for the update strategy diagram. Do you think it would be worth to make this compatibility window in between larger, for example, 3 versions?  I know we define an update strategy, but still.  I would like to keep potential problems here to a minimum. Or would that not be worth the effort?

File: lib/collection/src/collection.rs (hunk 3 above)
Comment (timvisee) [reply lib/collection/src/collection.rs:844]: Done in https://github.com/qdrant/qdrant/pull/2708

File: lib/collection/src/collection.rs (hunk 4 above)
Comment (timvisee) [reply lib/collection/src/collection.rs:949]: Done in https://github.com/qdrant/qdrant/pull/2708

File: src/tonic/api/points_internal_api.rs (hunk 5 above)
Comment (timvisee) [reply src/tonic/api/points_internal_api.rs:224]: Done in https://github.com/qdrant/qdrant/pull/2708

File: src/tonic/api/points_common.rs (hunk 6 above)
Comment (timvisee) [reply src/tonic/api/points_common.rs:761]: Done in https://github.com/qdrant/qdrant/pull/2708

File: lib/collection/src/shards/remote_shard.rs (hunk 8 above)
Comment (timvisee) [reply lib/collection/src/shards/remote_shard.rs:535]: Done in https://github.com/qdrant/qdrant/pull/2708

**Comment by timvisee** [comment]
@coszio I've tried to make some improvements in <https://github.com/qdrant/qdrant/pull/2708>. If those improvements seem legit, please merge it into this PR. :smiley:

**Comment by coszio** [comment]
Yes, of course we can make the window larger. Though we cannot merge your changes because it would try to communicate with an old version on the core search channel, which won't work. 

What I wanted with this plan was to get rid of duplicated code as soon as possible.

**Review by generall — APPROVED** [review]

**Pull request merged**